### DISCUSSION TITLE
Mutex<T> data wrapper

### DISCUSSION THREAD

**Discussion by ctigrisht** [dotnet/csharplang#6047] [Language Ideas]
It would be really nice if an object could be wrapped inside of a `Mutex<T>` in a way that when the value is accessed a lock occurs

```cs
// define the data
static ConcurrentDictionary<string, Mutex<MyClass>> MyData = new(); 

// access the data on a separate thread 
var mutexObj = MyData["key"];
mutexObj.Value.Property = "Some new data";
```

**Comment by jnm2** [comment]
How would the compiler know where the mutually-exclusive code starts and stops to achieve atomicity? It seems too implicit to be reliable for human readers, too.

**Reply by ctigrisht** [reply]
> How would the compiler know where the mutually-exclusive code starts and stops to achieve atomicity? It seems too implicit to be reliable for human readers, too.

Well I'm assuming the compiler would do some magic specifically for that type, but it does not seem implicit for human readers at all, it's just accessing the value in a thread safe manner

**Reply by jnm2** [reply]
What makes it thread-safe? Thread safety generally requires multiple things to happen atomically.

**Reply by jnm2** [reply]
Let's say you write this code:

```cs
var temp = mutexObj.Value.Counter;
temp += ExpensiveCalculation();
mutexObj.Value.Counter = temp;

SomethingExpensiveThatShouldNotBeInTheMutex();

var temp2 = mutexObj.Value.Counter;
temp2 += ExpensiveCalculation();
mutexObj.Value.Counter = temp2;
```

Is that thread-safe in your proposal? If so, how?

**Reply by ctigrisht** [reply]
The use case I envision from this would be having datasets of classes and whenever a write to one of those elements occur, it locks the item, ensuring that someone querying that data would have the latest version, it would also allow for multiple writes at the same time. when accessing the data you should not do things that will take long, it would be preferable to do the calculations before or after.
And if that expensive calculation does write back to the object, then it makes sense to lock the writes for that property on that object.

Basically it would only lock the fields/properties that are being written to at assignment time, this way it doesn't block for long

**Reply by jnm2** [reply]
> Basically it would only lock the fields/properties that are being written to at assignment time, this way it doesn't block for long

This does not constitute thread safety in the majority of cases. It's equivalent to letting copies of this code race against each other:

```cs
int temp;
lock (mutexObj) temp = mutexObj.Value.Counter;
temp++;
lock (mutexObj) mutexObj.Value.Counter = temp;
```

That's not thread-safe in the least.
